### PR TITLE
Implement bundle burn validation and serialization (new)

### PR DESCRIPTION
**Pull request by dmidem**
This pull request introduces handling of the 'burn' field of the bundle. The changes are split across two new modules: `burn_validation` and `burn_serialization`.

**Burn Validation**
The `burn_validation` module ensures the correctness of burn operations. Specifically, it validates that each asset to be burned is unique, non-native, and has a non-zero amount.

**Burn Serialization**
The `burn_serialization` module is responsible for reading and writing assets and burns from/to a stream.

Additionally, I have made modifications to the `arb_amount` function in `zcash_primitives/src/transaction/components/amount.rs` to generate amounts from the `1..MAX_MONEY` range instead of the `-MAX_MONEY..MAX_MONEY` range (to pass new validation checks for burn items, whch require amounts to be positive).

Furthermore, I have updated the test vectors for zip_0244 to accommodate the changes in the transaction structure. I have included an empty burn field in the test vectors to ensure alignment with the updated transaction format.

### CODE REVIEW

File: zcash_primitives/src/transaction/components/orchard.rs [hunk 1]
```diff
@@ -213,6 +219,7 @@ pub fn write_v5_bundle<W: Write>(
 
         writer.write_all(&[bundle.flags().to_byte()])?;
         writer.write_all(&bundle.value_balance().to_i64_le_bytes())?;
+        write_bundle_burn(&mut writer, bundle.burn())?;
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard.rs:222]: Let's inline this function, similar to two lines below.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        if !asset_set.insert(*asset) {
+            return Err(BurnError::DuplicateAsset);
+        }
+        if asset.is_native().into() {
+            return Err(BurnError::NativeAsset);
+        }
+        // FIXME: check for negative amounts?
+        if i64::from(amount) == 0 {
+            return Err(BurnError::ZeroAmount);
+        }
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:55]: Yes, and change the error to `NonPossitiveAmount`

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            BurnError::DuplicateAsset => write!(f, "Encountered a duplicate asset to burn."),
+            BurnError::NativeAsset => write!(f, "Cannot burn a native asset."),
+            BurnError::ZeroAmount => write!(f, "Cannot burn an asset with a zero amount."),
+        }
+    }
+}
+
+/// Validates burns for a bundle by ensuring each asset is unique, non-native, and has a non-zero value.
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:27]: Validates the burn values for a bundle by ensuring each asset is unique, non-native, and has a possitive value.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+///
+/// Each burn element is represented as a tuple of `AssetBase` and `Amount`, where `AssetBase` identifies
+/// the asset to be burned and `Amount` is the quantity to burn.
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:30]: replace with 
```
For burn, each asset is represented as a tuple of `AssetBase` and `Amount`.
```

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+///
+/// # Arguments
+///
+/// * `burns` - A vector of burns, where each burn is represented as a tuple of `AssetBase` and `Amount`.
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:34]: `burns` is not a good word.
```
* `burn` - A vector of assets, where each asset is represented as a tuple of `AssetBase` and `Amount`.
```

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+///
+/// # Errors
+///
+/// Returns a `BurnError` if:
+/// * Any asset in the list of burns is not unique (`BurnError::DuplicateAsset`).
+/// * Any asset in the list of burns is native (`BurnError::NativeAsset`).
+/// * Any asset in the list of burns has a zero amount (`BurnError::ZeroAmount`).
+pub fn validate_bundle_burn(bundle_burn: &Vec<(AssetBase, Amount)>) -> Result<(), BurnError> {
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:42]: `bundle_burn` -> `burn`

File: zcash_primitives/src/transaction/components/orchard/burn_serialization.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    use super::*;
+
+    use std::io::Cursor;
+
+    use crate::transaction::tests::create_test_asset;
+
+    use super::super::burn_validation::BurnError;
+
+    #[test]
+    fn test_read_write_bundle_burn_success() {
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_serialization.rs:70]: remove `test_` from all tests

File: zcash_primitives/src/transaction/tests.rs [hunk 8]
```diff
@@ -23,6 +23,18 @@ use super::{
...
+        keys::{IssuanceAuthorizingKey, IssuanceKey, IssuanceValidatingKey},
+        note::AssetBase,
+    };
+
+    let sk_iss = IssuanceKey::from_bytes([0u8; 32]).unwrap();
+    let isk: IssuanceAuthorizingKey = (&sk_iss).into();
+
+    AssetBase::derive(&IssuanceValidatingKey::from(&isk), asset_desc)
+}
+
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/tests.rs:37]: Insert new work at the bottom of the file, not at the top. (the entire new function)

File: zcash_primitives/src/transaction/tests.rs [hunk 9]
```diff
@@ -23,6 +23,18 @@ use super::{
 #[cfg(feature = "zfuture")]
 use super::components::tze;
 
+pub fn create_test_asset(asset_desc: &str) -> orchard::note::AssetBase {
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/tests.rs:26]: this function will be better utilized  if returns the burn tupple:
```rust
get_burn_tupple(asset_desc: &str, amount u64) -> (AssetBase, Amount) {
```

Comment (PaulLaux) [reply zcash_primitives/src/transaction/tests.rs:26]: please add a comment that the function is deterministic by using a static `sk_iss`.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs [hunk 10]
```diff
@@ -0,0 +1,118 @@
+use std::fmt;
+
+use orchard::note::AssetBase;
+
+use super::Amount;
+
+// FIXME: Consider making tuple (AssetBase, Amount) a new type.
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_validation.rs:7]: Think a tuple is simple enough for this.

File: zcash_primitives/src/transaction/components/orchard/burn_serialization.rs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        write_bundle_burn(&mut cursor, &bundle_burn).unwrap();
+
+        cursor.set_position(0);
+        let result = read_bundle_burn(&mut cursor).unwrap();
+
+        assert_eq!(result, bundle_burn);
+    }
+
+    // This test implementation covers only one failure case intentionally,
+    // as the other cases are already covered in the validate_bundle_burn tests.
```
Comment (PaulLaux) [inline zcash_primitives/src/transaction/components/orchard/burn_serialization.rs:91]: Either describe this test or not at all.
No need to discuss other tests.

**Review by PaulLaux — COMMENTED** [review]
Added some comments, plus we need a meaningful test vector in the python tests. https://github.com/QED-it/zcash-test-vectors/pull/10#discussion_r1246391923

File: zcash_primitives/src/transaction/components/orchard.rs (hunk 1 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard.rs]: Done.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs (hunk 2 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard/burn_validation.rs]: Done.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs (hunk 3 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard/burn_validation.rs]: Done.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs (hunk 4 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard/burn_validation.rs]: Done.

File: zcash_primitives/src/transaction/components/orchard/burn_validation.rs (hunk 5 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard/burn_validation.rs]: Done.

File: zcash_primitives/src/transaction/tests.rs (hunk 8 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/tests.rs]: Done.

File: zcash_primitives/src/transaction/tests.rs (hunk 9 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/tests.rs]: Done.

File: zcash_primitives/src/transaction/components/orchard/burn_serialization.rs (hunk 11 above)
Comment (dmidem) [reply zcash_primitives/src/transaction/components/orchard/burn_serialization.rs]: Done - removed this comment.

**Pull request closed without merging**